handout/http/search_with_switchmap.md

```diff
@@ -0,0 +1,91 @@
+# Enhancing Search with `switchMap`
+
+There is a problem with our previous implementation of incremental search.
+
+What if the server, for some reason, takes a very long time to respond to a particular query? If we use `flatMap`, we run the risk of getting results back from the server in the wrong order. Let's illustrate this with an example.
+
+## A Quick Example
+
+Consider a situation where we first type in the letters `ABC`, and suppose the string `ABC` is actually a special string where it will take the server a few extra seconds to reply.
+
+Meanwhile, after we paused for a bit (more than the debounce time), we decide to type in another letter (the letter X) and our app sends a request to the server for the string `ABCX`. Since `ABCX` is not considered a special string, the server replies very quickly and our app sets the suggestions for `ABCX`.
+
+A few seconds later, however, the server finally replies with the response for the `ABC` string, and our app receives that response and sets the search suggestions for `ABC`, overwriting the suggestions for the `ABCX` string, even though the request for that actually came afterwards.
+
+Here is a simple diagram to illustrate the issue:
+
+```
+// A1: Request for `ABC`
+// A2: Response for `ABC`
+// B1: Request for `ABCX`
+// B2: Response for `ABCX`
+
+--A1----------A2-->
+------B1--B2------>
+```
+
+You can see that A2 arrives after B2 even though the A1 request began first. So how do we get around this problem? We can use `switchMap`.
+
+## What is `switchMap`?
+
+SwitchMap is very similar to `flatMap`, but with a very important distinction. Any events to be merged into the trunk stream are ignored if a new event comes in. Here is a marble diagram showing the behavior of `switchMap`:
+
+![SwitchMap created by ReactiveX licensed under CC-3 (http://reactivex.io/documentation/operators/flatmap.html)](../images/switch-map.png)
+
+Just like `flatMap`, the red marble gets replaced with a red diamond and a subsequent red square. The interaction between the green and blue marble events are more interesting. Note that the green marble gets mapped to a green diamond immediately. And if enough time had passed, a green square would be pushed into the trunk stream but we do not see that here.
+
+Before the green square event is able to happen, a blue marble comes through and gets mapped to a blue diamond. What happened is that the green square is now ignored and do not get merged back into the trunk stream. The behavior of `switchMap` can be likened to a `flatMap` that "switches" to the more immediate incoming event and ignores all previously created event streams.
```

Comment:
I would add an extra paragraph for clarity:

In short, every time an event comes down the stream, `flatMap` will subscribe to a new observable without unsubscribing to any other observable created by a previous event. `switchMap` on the other hand will automatically unsubscribe to any previous observable when a new event comes down the stream.